core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java

```diff
@@ -0,0 +1,329 @@
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.Values;
+import org.apache.calcite.rel.logical.LogicalValues;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Collection of rules which simplify sections of query plan which are known to
+ * produce single row.
+ *
+ * <p>Conventionally, the way to represent a single row relational expression is
+ * with a {@link Values} that has one tuple.
+ *
+ * @see LogicalValues#createOneRow
+ */
+public abstract class SingleValueOptimizationRules {
+
+  public static final RelOptRule JOIN_LEFT_INSTANCE =
+      SingleValueOptimizationRules.JoinLeftSingleRuleConfig.DEFAULT.toRule();
+
+  public static final RelOptRule JOIN_RIGHT_INSTANCE =
+      SingleValueOptimizationRules.JoinRightSingleRuleConfig.DEFAULT.toRule();
+
+  public static final RelOptRule JOIN_LEFT_PROJECT_INSTANCE =
+      SingleValueOptimizationRules.JoinLeftSingleValueRuleWithExprConfig.DEFAULT.toRule();
+
+  public static final RelOptRule JOIN_RIGHT_PROJECT_INSTANCE =
+      SingleValueOptimizationRules.JoinRightSingleValueRuleWithExprConfig.DEFAULT.toRule();
+
+  /**
+   * Transformer class to transform a single value nodes on either side of the join.
+   * This transformer contains the common code for all the SingleValueJoin rules.
+   */
+  private static class SingleValueRelTransformer {
+
+    private final Join join;
+    private final RelNode relNode;
+    private final Predicate<Join> cannotTransform;
+    private final BiFunction<RexNode, List<RexNode>, List<RexNode>> litTransformer;
+    private final boolean valuesAsLeftChild;
+    private final List<RexNode> literals;
+
+    protected SingleValueRelTransformer(
+        Join join, List<RexNode> rexNodes, RelNode otherNode,
+        Predicate<Join> nonTransformable, boolean isValuesLeftChild,
+        BiFunction<RexNode, List<RexNode>, List<RexNode>> litTransformer) {
+      this.relNode = otherNode;
+      this.join = join;
+      this.cannotTransform = nonTransformable;
+      this.litTransformer = litTransformer;
+      this.valuesAsLeftChild = isValuesLeftChild;
+      this.literals = rexNodes;
+    }
+
+    public @Nullable RelNode transform(RelBuilder relBuilder) {
```

Comment:
I would give here a more elaborate comment. What happens is that a JOIN is converted into a FILTER where the filter predicate uses the join condition. I would in fact give a small example of a plan before and after.

Another very subtle point is that this works even if the VALUES contain NULL entries. This is because this generates code with comparisons to NULL, and comparing a value to NULL using `=` always returns `false`, so the result is an empty collection, as expected.

Reply:
Can you please add a comment in the code containing the second paragraph in my comment? I had to run in the debugger to understand why this happens.